bad-bits/lib/store.js

```diff
@@ -1,58 +1,106 @@
+import { getAllBadBitHashes } from '../test/util.js'
+
+const KV_VALUE_MAX_SIZE = 26214400
+const BAD_BITS_CID_LENGTH = 64
+const KV_SEGMENT_MAX_HASH_COUNT = Math.floor(
+  KV_VALUE_MAX_SIZE / (BAD_BITS_CID_LENGTH + 1) /* , */,
+)
+const MAX_TOTAL_CHANGES = 10_000
+const MAX_KV_BATCH_SIZE = 1_000
+
 /**
  * Updates the bad bits database with new hashes
  *
- * @param {Env} env - Environment containing database connection
+ * @param {{ KV: KVNamespace }} env - Environment containing database connection
  * @param {Set<string>} currentHashes - Set of current valid hashes from
  *   denylist
  * @param {string | null} etag - ETag for the current denylist
  */
 export async function updateBadBitsDatabase(env, currentHashes, etag) {
   const startedAt = Date.now()
-  const timestamp = new Date().toISOString()
-  const insertBadBitStmt = env.DB.prepare(
-    `
-    INSERT INTO bad_bits (hash, last_modified_at) VALUES (?, ?)
-    ON CONFLICT(hash) DO UPDATE SET last_modified_at = excluded.last_modified_at
-    `,
-  )
 
-  let updated = 0
-  const remainingHashes = Array.from(currentHashes)
-  while (remainingHashes.length > 0) {
-    // pop first N hashes from remainingHashes
-    const batchHashes = remainingHashes.splice(0, 10_000)
+  console.log('getting latest hashes')
+  const oldHashes = new Set(await getAllBadBitHashes(env))
+
+  const addedHashes = []
+  const removedHashes = []
 
-    await env.DB.batch(
-      batchHashes.map((hash) => insertBadBitStmt.bind(hash, timestamp)),
+  console.log('iterating current hashes')
+  for (const hash of currentHashes) {
+    if (!oldHashes.has(hash)) {
+      addedHashes.push(hash)
+    }
+    if (addedHashes.length >= MAX_TOTAL_CHANGES) {
+      break
+    }
+  }
+  for (const hash of oldHashes) {
+    if (!currentHashes.has(hash)) {
+      removedHashes.push(hash)
+    }
+    if (addedHashes.length + removedHashes.length >= MAX_TOTAL_CHANGES) {
+      break
+    }
+  }
+
+  console.log('writing added hashes')
+  for (let i = 0; i < Math.ceil(addedHashes.length / MAX_KV_BATCH_SIZE); i++) {
+    const batch = addedHashes.slice(
+      i * MAX_KV_BATCH_SIZE,
+      (i + 1) * MAX_KV_BATCH_SIZE - 1,
```

Comment:
```suggestion
      (i + 1) * MAX_KV_BATCH_SIZE ,
```

There's an off-by-one error here.